File: src/main/java/doom/despair/Player.java
package doom.despair;

import java.util.*;

public class Player {

    private final String displayName;
    private final UUID uuid;
    final List<Ship> ships;

    public Player(String displayName) {
        this.displayName = displayName;
        this.uuid = UUID.randomUUID();
        this.ships = new ArrayList<>();
    }

    public String getDisplayName() {
        return displayName;
    }

    public UUID getUuid() {
        return uuid;
    }

    public void addShip(Ship ship) {
        ships.add(ship);
    }

    public int getShipsRemaining() {
        return ships.size();
    }

    public boolean hasLost() {
        return getShipsRemaining() == 0;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Player other)) return false;
        return uuid.equals(other.uuid);
    }

    @Override
    public int hashCode() {
        return uuid.hashCode();
    }
}

File: src/main/java/doom/despair/Player.kt
package doom.despair

import java.util.*

class Player(@JvmField val displayName: String?) {
    @JvmField
    val uuid: UUID = UUID.randomUUID()

    @JvmField
    val ships: MutableList<Ship> = ArrayList<Ship>()

    fun addShip(ship: Ship) {
        ships.add(ship)
    }

    val shipsRemaining: Int
        get() = ships.size

    fun hasLost(): Boolean {
        return this.shipsRemaining == 0
    }

    override fun equals(other: Any?): Boolean {
        if (super.equals(other)) return true
        if (other !is Player) return false
        return uuid == other.uuid
    }

    override fun hashCode(): Int {
        return uuid.hashCode()
    }
}
